### ISSUE TITLE
TranslatableMixin system check does not recognise models.UniqueConstraint

### ISSUE THREAD

**Issue by enzedonline** [wagtail/wagtail#11098]
### Is your proposal related to a problem?

When making migrations, `TranslatableMixin` system check requires that a model has a `unique_together` constraint on `translation_key` and `locale`.

```bash
SystemCheckError: System check identified some issues:

ERRORS:
blog.BlogCategory: (wagtailcore.E003) blog.BlogCategory is missing a unique_together constraint for the translation key and locale fields
        HINT: Add ('translation_key', 'locale') to BlogCategory.Meta.unique_together
```

The current solution, add the `unique_together` definition to `Meta`:

```python
    class Meta:
        unique_together = ('translation_key', 'locale'), ('locale', 'slug')
```

However, Django docs state that [`models.UniqueConstraint`](https://docs.djangoproject.com/en/4.2/ref/models/constraints/#django.db.models.UniqueConstraint) should be used instead and that [`unique_together` may be deprecated in the future](https://docs.djangoproject.com/en/4.2/ref/models/options/#unique-together:~:text=UniqueConstraint%20provides%20more%20functionality%20than%20unique_together.%20unique_together%20may%20be%20deprecated%20in%20the%20future.).

In the above Meta, using the following will fail the system check:

```python
    class Meta:
        constraints = [
            models.UniqueConstraint(fields=['translation_key', 'locale'], name='unique_translation_key_locale'),
            models.UniqueConstraint(fields=['locale', 'slug'], name='unique_locale_slug'),
        ]
```

### Describe the solution you'd like

The system check is updated to recognise `models.UniqueConstraint` as an alternative to `unique_together`.
Perhaps even a warning level message added to echo Django doc's recommendation to use `models.UniqueConstraint` if `unique_together` has been used in the Meta definition.

**Comment by Temidayo32** [comment]
@enzedonline I see what you are saying. Let me look into this.

**Comment by Temidayo32** [comment]
So I have looked into it. There are several ways we can go about solving this:

- Throw a warning encouraging the use of UniqueConstraint if not used, while removing completely the need to check for `unique_together`.
```python

class TranslatableMixin(models.Model):
    translation_key = models.UUIDField(default=uuid.uuid4, editable=False)
    locale = models.ForeignKey(
        Locale, on_delete=models.PROTECT, related_name="+", editable=False
    )
    locale.wagtail_reference_index_ignore = True

    class Meta:
        abstract = True
        constraints = [
            models.UniqueConstraint(fields=['translation_key', 'locale'], name=f"unique_translation_key_locale_%(app_label)s_%(class)s")
        ]

    @classmethod
    def check(cls, **kwargs):
        errors = super().check(**kwargs)
        # Define the unique constraints you want to check for
        unique_constraints = [
            ('translation_key', 'locale'),
        ]

        # Check if each unique constraint exists
        for constraint_fields in unique_constraints:
            constraint_exists = any(
                isinstance(constraint, models.UniqueConstraint) and set(constraint.fields) == set(constraint_fields)
                for constraint in cls._meta.constraints
            )

            if not constraint_exists:
                errors.append(
                    checks.Warning(
                        "{}.{} is missing a UniqueConstraint for the fields: {}. Consider using models.UniqueConstraint.".format(
                            cls._meta.app_label, cls.__name__, constraint_fields
                        ),
                        hint="Add models.UniqueConstraint(fields={}, name='unique_{}') to {}.Meta.constraints.".format(
                            constraint_fields, "_".join(constraint_fields), cls.__name__
                        ),
                        obj=cls,
                        id="wagtailcore.W001",
                    )
                )

        return errors

```

**Comment by Temidayo32** [comment]
@enzedonline as a follow up on this. The second solution is more likely especially given Wagtail deprecation policy which I think would also apply here.

> Sometimes it is necessary for a feature release to deprecate features from previous releases. This will be noted in the “Upgrade considerations” section of the release notes.
> 
> When a feature is deprecated, it will continue to work in that feature release and the one after it, but will raise a warning. The feature will then be removed in the subsequent feature release. For example, a feature marked as deprecated in version 1.8 will continue to work in versions 1.8 and 1.9, and be dropped in version 1.10.

**Comment by enzedonline** [comment]
Yes, absolutely. unique_together would need to be phased out with a transition period where the system check tests for one or the other. The warning will give developers time to transition.

As per the initial suggestion
- UniqueConstraint be accepted by the SystemCheck as an alternative to unique_together (i.e. not replacing unique_together yet) - that is, the check is that one or the other is in place (and, for completeness, not both).
- Some notification (warning level or info if that exists) to let developers know that unique_together may be phased out and that UniqueConstraint should be used instead (with a link to the Django doc confirming this).

**Comment by Temidayo32** [comment]
@enzedonline Great. I linked a PR with the solution you offered. You can look through it and let me know what you think of it.

**Comment by laymonage** [comment]
I think we can consider this fixed in #11112 for now. I didn't add a warning/info if `unique_together` is used, because the default `TranslatableMixin.Meta` still uses it and updating it to use `UniqueConstraint` would force developers to generate migrations for their models.

We can take further steps once Django has officially deprecated `unique_together`, and follow what they'll do with [their own usage](https://github.com/django/django/blob/f7389c4b07ceeb036436e065898e411b247bca78/django/contrib/auth/models.py#L75) of it. When that time comes, we'll also have to update the docs for [`TranslatableMixin`](https://docs.wagtail.org/en/stable/reference/pages/model_reference.html#translatablemixin) and [bootstrapping models with it](https://docs.wagtail.org/en/stable/advanced_topics/i18n.html#translatable-snippets).
